File: src/features/catalog/uiState.ts
import { useSyncExternalStore } from "react";
import { ipc } from "@/ipc/client";
import type { LinksPlacementIpc, UiStateIpc } from "@/ipc/bindings";

/**
 * Shared read-modify-write cache for the global UI state (sort key + active
 * request + links placement).
 *
 * `ipc.appSettingsSet` REPLACES the entire persisted `UiStateIpc` — it is not a merge.
 * Multiple components persist different fields (SidebarShell → sort_key,
 * WorkflowApp → active_request, Settings → links_placement); funnelling every write
 * through this single cache lets `patchUiState` send the full merged object so the
 * writers never clobber each other.
 *
 * Writes also notify subscribers so reactive readers (e.g. `useLinksPlacement`)
 * re-render open panels the moment the setting changes.
 */
const DEFAULTS: UiStateIpc = { sort_key: null, active_request: null, links_placement: "strip" };
let cache: UiStateIpc = DEFAULTS;

const listeners = new Set<() => void>();
function emit(): void {
  for (const fn of listeners) fn();
}

/** Load the persisted UI state from the backend and seed the cache. */
export async function loadUiState(): Promise<UiStateIpc> {
  cache = await ipc.appSettingsGet();
  emit();
  return cache;
}

/** The last-known UI state (cache). Synchronous; call `loadUiState` first to populate. */
export function readUiState(): UiStateIpc {
  return cache;
}

/** Merge `patch` into the cache and persist the FULL merged object (no clobber). */
export async function patchUiState(patch: Partial<UiStateIpc>): Promise<void> {
  cache = { ...cache, ...patch };
  emit();
  await ipc.appSettingsSet(cache);
}

/** Subscribe to cache changes (load + patch). Returns an unsubscribe fn. */
export function subscribeUiState(fn: () => void): () => void {
  listeners.add(fn);
  return () => listeners.delete(fn);
}

/** Reactive read of the collection-links placement preference. Re-renders on change.
 *  The field is serde-optional over IPC (older settings files omit it), so absence
 *  reads as the `strip` default. */
export function useLinksPlacement(): LinksPlacementIpc {
  return useSyncExternalStore(subscribeUiState, () => cache.links_placement ?? "strip");
}

/** Test-only: reset the module-level cache between tests. */
export function resetUiState(): void {
  cache = DEFAULTS;
  emit();
}

File: src-tauri/src/commands/ui_state.rs
//! App-settings IPC commands: read/write the persisted [`UiStateIpc`] envelope
//! (sort key + active request). Thin wrappers over
//! [`AppState::ui_state_store`] — see [`crate::commands::collection`] for the
//! `#[tauri::command]` convention.

use tauri::State;

use crate::ipc::error::IpcError;
use crate::ipc::ui_state::UiStateIpc;
use crate::state::AppState;

#[tauri::command]
#[specta::specta]
pub async fn app_settings_get(state: State<'_, AppState>) -> Result<UiStateIpc, IpcError> {
    Ok(UiStateIpc::from_core(state.ui_state_store.get()))
}

/// Replaces the entire persisted UI state — callers send the complete object, not a partial patch.
#[tauri::command]
#[specta::specta]
pub async fn app_settings_set(state: State<'_, AppState>, patch: UiStateIpc) -> Result<(), IpcError> {
    state.ui_state_store.set(patch.into_core()).map_err(IpcError::from)
}
